core/libs/shared/src/file_metadata.rs

```diff
@@ -0,0 +1,232 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::str::FromStr;
+
+use libsecp256k1::PublicKey;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::access_info::{EncryptedFolderAccessKey, UserAccessInfo};
+use crate::account::{Account, Username};
+use crate::clock::get_time;
+use crate::crypto::AESKey;
+use crate::file_like::FileLike;
+use crate::secret_filename::SecretFileName;
+use crate::signed_file::SignedFile;
+use crate::{pubkey, symkey, SharedResult};
+
+pub type DocumentHmac = [u8; 32];
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct FileMetadata {
+    pub id: Uuid,
+    pub file_type: FileType,
+    pub parent: Uuid,
+    pub name: SecretFileName,
+    pub owner: Owner,
+    pub is_deleted: bool,
+    pub document_hmac: Option<DocumentHmac>,
+    pub user_access_keys: HashMap<Username, UserAccessInfo>,
+    pub folder_access_keys: EncryptedFolderAccessKey,
+}
+
+impl FileMetadata {
+    pub fn create_root(account: &Account) -> SharedResult<Self> {
+        let id = Uuid::new_v4();
+        let key = symkey::generate_key();
+        let pub_key = account.public_key();
+
+        Ok(FileMetadata {
+            id,
+            file_type: FileType::Folder,
+            parent: id,
+            name: SecretFileName::from_str(&account.username, &key, &key)?,
+            owner: Owner(pub_key),
+            is_deleted: false,
+            document_hmac: None,
+            user_access_keys: UserAccessInfo::encrypt(account, &pub_key, &key)?,
+            folder_access_keys: symkey::encrypt(&key, &key)?,
+        })
+    }
+
+    pub fn create(
+        pub_key: &PublicKey, parent: Uuid, parent_key: &AESKey, name: &str, file_type: FileType,
+    ) -> SharedResult<Self> {
+        let id = Uuid::new_v4();
+        let key = symkey::generate_key();
+
+        Ok(FileMetadata {
+            id,
+            file_type,
+            parent,
+            name: SecretFileName::from_str(name, &key, parent_key)?,
+            owner: Owner(*pub_key),
+            is_deleted: false,
+            document_hmac: None,
+            user_access_keys: Default::default(),
+            folder_access_keys: symkey::encrypt(parent_key, &key)?,
+        })
+    }
+
+    pub fn sign(self, account: &Account) -> SharedResult<SignedFile> {
+        pubkey::sign(&account.private_key, self, get_time)
+    }
+}
+
+impl PartialEq for FileMetadata {
```

Comment:
Why can't this be derived? If there's a reason, I think we should state it in a doc comment right here.

Reply:
https://github.com/lockbook/lockbook/issues/1251